x/farming/keeper/proposal_handler.go

```diff
@@ -1,54 +1,174 @@
 package keeper
 
 import (
-	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 
 	"github.com/tendermint/farming/x/farming/types"
 )
 
-// HandlePublicPlanProposal is a handler for executing a fixed amount plan creation proposal.
-func HandlePublicPlanProposal(ctx sdk.Context, k Keeper, plansAny []*codectypes.Any) error {
-	plans, err := types.UnpackPlans(plansAny)
-	if err != nil {
-		return err
+// HandlePublicPlanProposal is a handler for executing a public plan creation proposal.
+func HandlePublicPlanProposal(ctx sdk.Context, k Keeper, proposal *types.PublicPlanProposal) error {
+	if err := proposal.ValidateBasic(); err != nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
 	}
 
-	for _, plan := range plans {
-		switch p := plan.(type) {
-		case *types.FixedAmountPlan:
+	switch {
+	case proposal.AddRequestProposals != nil:
+		if err := k.AddPublicPlanProposal(ctx, proposal.AddRequestProposals); err != nil {
+			return err
+		}
+	case proposal.UpdateRequestProposals != nil:
+		if err := k.UpdatePublicPlanProposal(ctx, proposal.UpdateRequestProposals); err != nil {
+			return err
+		}
+	case proposal.DeleteRequestProposals != nil:
+		if err := k.DeletePublicPlanProposal(ctx, proposal.DeleteRequestProposals); err != nil {
+			return err
+		}
+	default:
+		return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unexpected public plan proposal %s", proposal.String())
+	}
+
+	return nil
+}
+
+// AddPublicPlanProposal adds a new public plan once the governance proposal is passed.
+func (k Keeper) AddPublicPlanProposal(ctx sdk.Context, proposals []*types.AddRequestProposal) error {
+	for _, p := range proposals {
+		farmingPoolAddrAcc, err := sdk.AccAddressFromBech32(p.GetFarmingPoolAddress())
+		if err != nil {
+			return err
+		}
+
+		if !p.EpochAmount.IsZero() && !p.EpochAmount.IsAnyNegative() {
 			msg := types.NewMsgCreateFixedAmountPlan(
-				p.GetFarmingPoolAddress(),
+				p.GetName(),
+				farmingPoolAddrAcc,
 				p.GetStakingCoinWeights(),
 				p.GetStartTime(),
 				p.GetEndTime(),
 				p.EpochAmount,
 			)
 
-			fixedPlan := k.CreateFixedAmountPlan(ctx, msg, types.PlanTypePublic)
+			plan, err := k.CreateFixedAmountPlan(ctx, msg, types.PlanTypePublic)
+			if err != nil {
+				return err
+			}
 
 			logger := k.Logger(ctx)
-			logger.Info("created public fixed amount plan", "fixed_amount_plan", fixedPlan)
+			logger.Info("created public fixed amount plan", "fixed_amount_plan", plan)
+		}
 
-		case *types.RatioPlan:
+		if !p.EpochRatio.IsZero() && !p.EpochRatio.IsNil() && !p.EpochRatio.IsNegative() {
```

Comment:
This time we must use `if-else if` or `switch-case` because only one type of plan should be created.

Reply:
Current logic creates only one type of plan as far as I test this. Are you suggesting to use `if-else if` just to make sure if doesn't go through the next `if` when `epoch_amount` exists?

Reply:
Yeah, that's why I suggested.